### PR TITLE
Federation kubectl for clusters

### PR DESCRIPTION
**Pull request by jianhuiz**
add federation/clusters resource to kubectl
#23653, requires #23847

---

This change is [Reviewable](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/24016)

### CODE REVIEW

**Comment by ihmccreery** [comment]
@nikhiljindal I'm just passing these off to you because #23847 was assigned to you.

**Comment by madhusudancs** [comment]
I will be happy to review this one. Please let me know when you are ready.

cc @kubernetes/sig-cluster-federation

**Comment by nikhiljindal** [comment]
@jianhuiz Is this still WIP or ready to review?

File: pkg/api/validation/schema.go [hunk 1]
```diff
@@ -219,8 +219,8 @@ func (s *SwaggerSchema) ValidateObject(obj interface{}, fieldName, typeName stri
 	return allErrs
 }
 
-// This matches type name in the swagger spec, such as "v1.Binding".
-var versionRegexp = regexp.MustCompile(`^v.+\..*`)
+// This matches type name in the swagger spec, such as "v1.Binding", "unversioned.ServerAddressByClientCIDR.
+var versionRegexp = regexp.MustCompile(`^(v.+|unversioned)\..*`)
```
Comment (madhusudancs) [inline pkg/api/validation/schema.go]: Why this change here?

File: pkg/kubectl/cmd/util/clientcache.go [hunk 2]
```diff
@@ -17,30 +17,36 @@ limitations under the License.
 package util
 
 import (
+	fed_client_1_3 "k8s.io/kubernetes/federation/client/clientset_generated/release_1_3"
```
Comment (madhusudancs) [inline pkg/kubectl/cmd/util/clientcache.go]: @caesarxuchao Just wanted you to take a look at this. Do you think this is the right "new" way to import the clients? Particularly, should the import name have the version as well?

File: pkg/kubectl/cmd/util/clientcache.go [hunk 3]
```diff
@@ -125,3 +131,71 @@ func (c *ClientCache) ClientForVersion(version *unversioned.GroupVersion) (*clie
 
 	return kubeclient, nil
 }
+
+func (c *ClientCache) KubeClientSetForVersion(version *unversioned.GroupVersion) (kube_client_1_3.Interface, error) {
```
Comment (madhusudancs) [inline pkg/kubectl/cmd/util/clientcache.go]: Where is this function used right now? Will it be used in your future commits?

File: contrib/completions/bash/kubectl [hunk 4]
```diff
@@ -431,9 +432,11 @@ _kubectl_get()
     must_have_one_noun+=("secret")
     must_have_one_noun+=("service")
     must_have_one_noun+=("serviceaccount")
+    must_have_one_noun+=("subreplicaset")
```
Comment (madhusudancs) [inline contrib/completions/bash/kubectl]: Assuming you are going to remove the subReplicaSet parts, this commit generally looks good. I would like @nikhiljindal to make a final pass for client side changes once this PR is in its final shape (not WIP).

File: pkg/kubectl/cmd/util/clientcache.go (hunk 3 above)
Comment (jianhuiz) [inline pkg/kubectl/cmd/util/clientcache.go]: It was intent to be used in describers which may need both the kube client and the fed client.
Since we will have only clusters now, I will remove it now.

File: pkg/api/validation/schema.go (hunk 1 above)
Comment (nikhiljindal) [inline pkg/api/validation/schema.go]: I talked to @lavalamp and he suggested that its better to copy ServerAddressByClientCIDR in federation types.go rather than using an unversioned struct there.
@jianhuiz Can you please send a PR for that.
We wont require this change then.

File: pkg/kubectl/cmd/util/clientcache.go (hunk 2 above)
Comment (caesarxuchao) [inline pkg/kubectl/cmd/util/clientcache.go]: I think the code you referred is out-dated, rigth? I agreed with @jianhuiz yesterday that we would use the internalversion clientset for now.

File: pkg/api/validation/schema.go (hunk 1 above)
Comment (jianhuiz) [inline pkg/api/validation/schema.go]: new PR at #24943

File: pkg/kubectl/cmd/util/clientcache.go (hunk 2 above)
Comment (madhusudancs) [inline pkg/kubectl/cmd/util/clientcache.go]: @caesarxuchao yeah, I think you are right. Sorry, I wasn't in the room when that conversion happened.

**Comment by jianhuiz** [comment]
1. `pkg/kubectl/describe.go` need to import `federation/client/clientset_generated/federation_internalclientset` to get cluster objects.
2. as an convention, `federation/client/clientset_generated/federation_internalclientset/import_known_versions.go` install federation group types
3. deployment controller (`pkg/controller/deployment/deployment_controller.go`) imports `pkg/kubectl` for an annotation const (`LastAppliedConfigAnnotation`) #20678
4. master uees deployment controller
5. kube-apiserver uses master

Result: federation/v1alpha1 registered to kube-apiserver causes the generated `docs/admin/kube-apiserver.md` contains `federation/v1alpha1` in `--storage-versions`

Resolution:
1. move describe.go to a sub package so import kubectl will not import federation client package
2. do not use import_known_versions.go trick. Is depending on package import side effect a good solution?
3. move annotion const define to deployment controler (reverse the dependency)

@nikhiljindal @madhusudancs @quinton-hoole  @janetkuo

**Comment by nikhiljindal** [comment]
So the dependency is:
`kube-apiserver -> pkg/master -> deployment-controller -> kubectl -> federated_clientset -> federated_clientset/import_known_versions.go -> federation/v1alpha1`

Where does master use deployment_controller?

How about moving LastAppliedConfigAnnotation to a separate package so that there is no dependency between deployment_controller and kubectl?

cc @caesarxuchao Is there a plan to stop requiring import_known_versions  in the generated client: federation/client/clientset_generated/federation_internalclientset/import_known_versions.go ?

**Comment by jianhuiz** [comment]
@nikhiljindal I was wrong about "master use deployment_controller", it's that "cmd/genkubedocs/gen_kube_docs.go imports kube-controller-manager/app and kube-apiserver/app at the same time". updated the original post.

**Comment by caesarxuchao** [comment]
> How about moving LastAppliedConfigAnnotation to a separate package so that there is no dependency between deployment_controller and kubectl?

I think this is the way to go. The deployment controller shouldn't depend on the entire kubectl package.

> @caesarxuchao Is there a plan to stop requiring import_known_versions in the generated client: federation/client/clientset_generated/federation_internalclientset/import_known_versions.go ?

No, import_known_versions.go is our compile time switch of APIs.

**Comment by jianhuiz** [comment]
@nikhiljindal @caesarxuchao In regarding "moving LastAppliedConfigAnnotation to a separate package", which package do you suggest? There are quite a few other annotations in kubectl, do they need to be moved  as well?

**Comment by nikhiljindal** [comment]
I discussed this with @lavalamp and `pkg/api/annotations` is what we decided.
`pkg/api` and `pkg/apis` contain objects that provide contract between the client and server. This annotation is also doing that.
You can name the package 'annotations' and add a comment in pkg/api/annotations/doc.go that all annotations shared between the server and client should go here.
For now, it will only have the LastAppliedConfigAnnotation annotation.

File: pkg/kubectl/apply.go [hunk 5]
```diff
@@ -28,23 +29,19 @@ type debugError interface {
...
 
-// LastAppliedConfigAnnotation is the annotation used to store the previous
-// configuration of a resource for use in a three way diff by UpdateApplyAnnotation.
-const LastAppliedConfigAnnotation = kubectlAnnotationPrefix + "last-applied-configuration"
-
 // GetOriginalConfiguration retrieves the original configuration of the object
 // from the annotation, or nil if no annotation was found.
 func GetOriginalConfiguration(info *resource.Info) ([]byte, error) {
-	annotations, err := info.Mapping.MetadataAccessor.Annotations(info.Object)
+	annots, err := info.Mapping.MetadataAccessor.Annotations(info.Object)
```
Comment (brendandburns) [inline pkg/kubectl/apply.go]: why did you change this?  its less descriptive than the original name.

Comment (brendandburns) [inline pkg/kubectl/apply.go]: ah, nm, its the package import.  *sigh*

File: pkg/kubectl/cmd/util/clientcache.go [hunk 6]
```diff
@@ -125,3 +128,41 @@ func (c *ClientCache) ClientForVersion(version *unversioned.GroupVersion) (*clie
 
 	return kubeclient, nil
 }
+
+func (c *ClientCache) FederationClientSetForVersion(version *unversioned.GroupVersion) (fed_clientset.Interface, error) {
+	if version != nil {
+		if clientSet, ok := c.fedClientSets[*version]; ok {
```
Comment (brendandburns) [inline pkg/kubectl/cmd/util/clientcache.go]: prefer `found` to `ok` for map lookup

File: pkg/kubectl/describe.go [hunk 7]
```diff
@@ -1937,6 +1939,63 @@ func describeConfigMap(configMap *api.ConfigMap) (string, error) {
...
+	cluster, err := d.Federation().Clusters().Get(name)
+	if err != nil {
+		return "", err
+	}
+	return describeCluster(cluster)
+}
+
+func describeCluster(cluster *federation.Cluster) (string, error) {
+	return tabbedString(func(out io.Writer) error {
+		fmt.Fprintf(out, "Nmae:\t%s\n", cluster.Name)
```
Comment (brendandburns) [inline pkg/kubectl/describe.go]: s/Nmae/Name/

**Comment by jianhuiz** [comment]
fixed and rebased

**Comment by jianhuiz** [comment]
rebased with #23847 merged

**Comment by nikhiljindal** [comment]
LGTM, thanks!

We also want `kubectl get clusters` to come up in `kubectl help`, but it should only show up when kubectl is talking to a federation apiserver.
We can do that by configuring the output based on discovery API from apiserver rather than being a hardcoded list in kubectl.

**Comment by jianhuiz** [comment]
rebased

File: pkg/kubectl/cmd/util/factory.go [hunk 8]
```diff
@@ -346,6 +351,15 @@ func NewFactory(optionalClientConfig clientcmd.ClientConfig) *Factory {
 		},
 		Describer: func(mapping *meta.RESTMapping) (kubectl.Describer, error) {
 			mappingVersion := mapping.GroupVersionKind.GroupVersion()
+			if mapping.GroupVersionKind.Group == federation.GroupName {
+				fedClientSet, err := clients.FederationClientSetForVersion(&mappingVersion)
+				if err != nil {
+					return nil, err
+				}
+				if mapping.GroupVersionKind.Kind == "Cluster" {
+					return &kubectl.ClusterDescriber{fedClientSet}, nil
```
Comment (nikhiljindal) [inline pkg/kubectl/cmd/util/factory.go]: You need to change this to `&kubectl.ClusterDescriber{Interface: fedClientSet}`
hack/verify-govet is failing with:
```
pkg/kubectl/cmd/util/factory.go:360: k8s.io/kubernetes/pkg/kubectl.ClusterDescriber composite literal uses unkeyed fields
```

File: pkg/kubectl/describe_test.go [hunk 9]
```diff
@@ -522,6 +524,40 @@ func TestDescribeDeployment(t *testing.T) {
...
+			},
+		},
+		Status: federation.ClusterStatus{
+			Conditions: []federation.ClusterCondition{
+				{Type: federation.ClusterReady, Status: api.ConditionTrue},
+			},
+		},
+	}
+	fake := fed_fake.NewSimpleClientset(&cluster)
+	d := ClusterDescriber{fake}
```
Comment (nikhiljindal) [inline pkg/kubectl/describe_test.go]: similar fix is required here

File: pkg/kubectl/cmd/util/factory.go [hunk 10]
```diff
@@ -346,6 +351,15 @@ func NewFactory(optionalClientConfig clientcmd.ClientConfig) *Factory {
 		},
 		Describer: func(mapping *meta.RESTMapping) (kubectl.Describer, error) {
 			mappingVersion := mapping.GroupVersionKind.GroupVersion()
+			if mapping.GroupVersionKind.Group == federation.GroupName {
```
Comment (nikhiljindal) [inline pkg/kubectl/cmd/util/factory.go]: Another way to implement this is to have a way for kubectl to detect that it is talking to federation apiserver (using discovery api) and then always use the federation clientset and FederationDescriber.
That way, we can implement different federation describer for `kubectl describe services` as well.

Comment (nikhiljindal) [inline pkg/kubectl/cmd/util/factory.go]: To clarify, I dont want you to change that in this PR. Consider the above as a thought that we can discuss more.

File: pkg/kubectl/cmd/util/factory.go (hunk 8 above)
Comment (jianhuiz) [inline pkg/kubectl/cmd/util/factory.go]: fixed

File: pkg/kubectl/describe_test.go (hunk 9 above)
Comment (jianhuiz) [inline pkg/kubectl/describe_test.go]: fixed

File: pkg/kubectl/cmd/util/factory.go (hunk 10 above)
Comment (jianhuiz) [inline pkg/kubectl/cmd/util/factory.go]: ok, maybe put it in #25592

**Comment by nikhiljindal** [comment]
@k8s-bot node e2e test this issue: https://github.com/kubernetes/kubernetes/issues/25375

**Comment by nikhiljindal** [comment]
thanks for the fixes.
Squash the last commit?

**Comment by jianhuiz** [comment]
fixed and rebased

**Comment by nikhiljindal** [comment]
@k8s-bot node e2e test this issue: https://github.com/kubernetes/kubernetes/issues/25713

**Comment by nikhiljindal** [comment]
All tests pass!
LGTM, thanks

**Comment by bgrant0607** [comment]
cc @kubernetes/kubectl
